Train resnet18 on cropped dogs. Mian idea is that there are a lot of non-useful information in imagewoof photos. I cropped photos (check "crop and mask.ipynb") and train resnet18 on new data.

In [1]:
import os 
import pandas as pd
import numpy as np
from tqdm.notebook import tqdm
from PIL import Image, ImageDraw

import torch
from torch import nn
import torchvision
from torch.nn import functional as F

import matplotlib.pyplot as plt

from utils.data import WoofDataset
from utils.train import classic_train

from torch.utils.data import DataLoader

In [2]:
torch.manual_seed(88)
np.random.seed(88)

device =  torch.device(('cuda:0' if torch.cuda.is_available() else 'cpu'))  # включаем поддержку GPU
device

device(type='cuda', index=0)

In [38]:
train_set = WoofDataset('data/woof_cut/train/boxed', (256, 256))
val_set = WoofDataset('data/woof_cut/val/boxed', (256, 256))

In [39]:
BATCH_SIZE = 50
NUM_WORKERS = 4
NAME_TEST = 'boxed_resnet18'
NUM_TEST = 0

if not os.path.isdir('weights/' + NAME_TEST):
    os.mkdir('weights/' + NAME_TEST)

train_loader = DataLoader(train_set, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS)
val_loader = DataLoader(val_set, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS)

In [40]:
net = torchvision.models.resnet18(pretrained=True)
net.fc = nn.Linear(512, 10)

net = net.to(device)

In [41]:
loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(net.parameters(), lr = 0.0001, amsgrad=True)

In [42]:
result = classic_train(net, 180, device, loss_fn, optimizer, train_loader, val_loader, num_epoch=10, NUM_TEST=NUM_TEST, NAME_TEST=NAME_TEST, begin=0)

-- EPOCH 0 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...


/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:3335: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:161: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


Epoch 0 Iteration 0 
Train loss = nan 
Val loss = 2.5305829289593276 
Accuracy = 0.1556978233034571 
F-score = 0.07334361527773131 

-- EPOCH 1 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 1 Iteration 180 
Train loss = 0.5385147878279288 
Val loss = 0.3658845909411394 
Accuracy = 0.8852752880921895 
F-score = 0.8857659526697065 

-- EPOCH 2 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 2 Iteration 360 
Train loss = 0.15916698982732164 
Val loss = 0.3447969636490828 
Accuracy = 0.8942381562099871 
F-score = 0.8935544218954973 

-- EPOCH 3 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 3 Iteration 540 
Train loss = 0.05720381592400372 
Val loss = 0.3412337280526946 
Accuracy = 0.8985915492957747 
F-score = 0.8978208172258694 

-- EPOCH 4 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 4 Iteration 720 
Train loss = 0.019395307697252267 
Val loss = 0.32558071877382977 
Accuracy = 0.9062740076824584 
F-score = 0.9059706135360293 

-- EPOCH 5 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 5 Iteration 900 
Train loss = 0.007200236240815785 
Val loss = 0.3198952218399772 
Accuracy = 0.9096030729833546 
F-score = 0.9092820844862382 

-- EPOCH 6 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 6 Iteration 1080 
Train loss = 0.005122060445784074 
Val loss = 0.32193323842520954 
Accuracy = 0.9119078104993598 
F-score = 0.9114556767525114 

-- EPOCH 7 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 7 Iteration 1260 
Train loss = 0.0036216147418599576 
Val loss = 0.3273447488517135 
Accuracy = 0.9134443021766966 
F-score = 0.9131841963089814 

-- EPOCH 8 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 8 Iteration 1440 
Train loss = 0.002307614166380113 
Val loss = 0.3306073695656997 
Accuracy = 0.912932138284251 
F-score = 0.9125139530657698 

-- EPOCH 9 ---------------------


  0%|          | 0/180 [00:00<?, ?it/s]

Validating...
Epoch 9 Iteration 1620 
Train loss = 0.0020095932313577377 
Val loss = 0.3286969609487849 
Accuracy = 0.9137003841229193 
F-score = 0.9134338483521565 



Maximum Accuracy is 91.3%
It is much less than classic method accuracy. This is a bit strange. I will inverstigate that problem.